// http-handler/index.js
if (request == null) var request = require('request');

var userPass = "cmput402:qpskcnvb";
var mapQuestAPIKey = "Fmjtd%7Cluu82qu1nh%2Cax%3Do5-94y2dy";
var retrieveAffiliationsJSON = function (successCb, failureCb, args) {
  var query = args['query'];

      // Set the headers
    var headers = {
        'Content-type' : 'application/x-www-form-urlencoded',
        'Authorization': 'Basic ' + new Buffer(userPass).toString('base64')
    }

    // Configure the request
    var options = {
        url: "http://199.116.235.47/centralrepo/v1/pull_affiliations",
        headers: headers,
        form: query
    }
    console.log(query)
    // Start the request
    request.post(options, function (error, response, body) {
         // Print out the response body
         successCb(JSON.parse(body));
    })
}

var retrieveDocumentsByTitleJSON = function (successCb, failureCb, args) {
  var keyword = args['keyword'];
    console.log(keyword)
      // Set the headers
    var headers = {
        'Content-type' : 'application/json',
        'Authorization': 'Basic ' + new Buffer(userPass).toString('base64')
    }

    // Configure the request
    var options = {
        url: "http://129.128.184.88/api/getDocumentsByTitle/" + keyword,
        headers: headers,
    }
    console.log(options.url)
    // Start the request
    request.get(options, function (error, response, body) {
         // Print out the response body
         successCb(JSON.parse(body));
    })
}

var retrieveCoordByLocationJSON = function (successCb, failureCb, args) {
  var locations = JSON.stringify(args['locations']);
  // Configure the request
  var options = {
    url:  "http://open.mapquestapi.com/geocoding/v1/batch?key=" + mapQuestAPIKey + 
    "&outFormat=json&inFormat=json&json=" + locations
  }

  request.get(options, function (error, response, body) {
     // Print out the response body
     successCb(JSON.parse(body));
  })
}

var retrieveInstitutionJSON = function (successCb, failureCb, args) {
  var keyword = args['keyword'];
    console.log(keyword)
      // Set the headers
    var headers = {
        'Content-type' : 'application/json',
        'Authorization': 'Basic ' + new Buffer(userPass).toString('base64')
    }

    // Configure the request
    var options = {
        url: "http://129.128.184.88/api/getBasicInstitutionInfo/start/0/count/200/" + keyword,
        headers: headers,
    }
    console.log(options.url)
    // Start the request
    request.get(options, function (error, response, body) {
         // Print out the response body
         successCb(JSON.parse(body));
    })
}


exports.retrieveInstitutionJSON = retrieveInstitutionJSON;
exports.retrieveCoordByLocationJSON = retrieveCoordByLocationJSON;
exports.retrieveDocumentsByTitleJSON = retrieveDocumentsByTitleJSON;
exports.retrieveAffiliationsJSON = retrieveAffiliationsJSON;
